### DISCUSSION TITLE
Bugs in setup??

### DISCUSSION THREAD

**Discussion by gwww** [folke/lua-dev.nvim#115] [General]
I have run into a weird bug where when I startup nvim it complains about `Undefined global: `vim``. I've spent the last 4 hours putting print statements all over the neodev code and the lspconfig code and I've narrowed it down to a couple of lines in neodev that seem to be causing the problem.

This [line](https://github.com/folke/neodev.nvim/blob/main/lua/neodev/lsp.lua#L10):
```
  lsputil.on_setup = hook(lsputil.on_setup, function(config)
```
seems odd. Feels like it should be:
```
  lsputil.on_setup = hook(lsputil.setup, function(config)
```

This is followed by the next line. It does not pass through this line because `opts` is nil (for reasons unknown to be me yet):
```
    if opts.setup_jsonls and config.name == "jsonls" then
```
I think adding a guard seems right, so this:
```
    if opts and opts.setup_jsonls and config.name == "jsonls" then
```

When I make both those changes in I no longer get my error. But, I don't know the bigger context of how this code was designed. My analysis could be totally wrong.

**Comment by folke** [comment]
That `opts` can't be `nil`, so not sure how you get that to be `nil` in your setup. What does your full neodev setup look like?

Also, your `"analysis"` here above, is wrong. That `on_setup` should be `on_setup`, etc.

**Reply by gwww** [reply]
Setup is a bit complicated. I'm using AstroNvim. What is complicated is that the method for loading plugins changed overnight.

When the problem occurs for me I was running https://github.com/AstroNvim/AstroNvim/tree/baaf407f12ce9e5bcd4e17bd2c48c1ce65f6b7c1 (which is one commit back from HEAD, as I write this).

The plugin loading changed last night. I'm now retesting with the latest to try and reproduce the problem.

**Reply by gwww** [reply]
One more thing. The complicated bit is that the loading of neodev works up until I do an update of AstroNvim. So, with a freshly cloned AstroNvim and freshly cloned plugins, everything work. As soon as I run the AstroNvim update function (https://github.com/AstroNvim/AstroNvim/blob/baaf407f12ce9e5bcd4e17bd2c48c1ce65f6b7c1/lua/core/utils/updater.lua#L140) neodev stops loading.

**Comment by andrewcrook** [comment]
> Undefined global: vim

I don't think that an issue with this plugin, I was getting that before installing.
lua_ls has the following setting but I cannot get it to work for some reason.

```
   diagnostics = {
                    enable = true,
                    -- Get the language server to recognize the `vim` global
                    globals = { ‘vim' },
                },
```
